### DISCUSSION TITLE
Clarification on geometry requirements

### DISCUSSION THREAD

**Discussion by Dunge** [usdot-jpo-ode/jpo-wzdx#193] [Q&A] [answered]
## Summary

The current geometry requirements are vague. It only states it needs to conform GeoJSON specifications, and is described as:

> The geometry of the road event. The Geometry object's type property MUST be LineString or MultiPoint. LineString allows specifying the entire road event path and should be preferred. MultiPoint should be used when only the start and end coordinates are known.

Questions:
* Can an event be composed of a single point? i.e. we know the start, but not the end. Can we push the same coordinate twice in MultiPoint mode?
* Does a LineString event needs to follow the road curvatures or can be an approximation?

 ## Motivation
I will expend on the second question.
When generating a LineString, we have a few options. Especially in the context of automating WZDx generation based on device GPS positions.

1. Draw on the map (the best case, require manual entry)
2. Automate and receive the path via mapping services like for example Google Maps based on the GPS waypoints.
3. Automate and generate lines going from one GPS position to the other without following the road curvature.

We used the **second method** for a while, but found it to be too unstable since it can lead to incorrect information in multiple situations:
* GPS is imprecise and appear in the other lane direction, or in a lane that goes near an highway exit.
* Mapping service detect a road closure and returns a path trying to avoid it by taking a detour.
* Mapping service is not aware of a temporary road placement and still use the normal road position.

This lead to obviously incorrect results like this
![image](https://user-images.githubusercontent.com/6548184/128784092-1f138b20-10f2-498d-9400-63ca6dd4e3cb.png)

So this lead us to change to a simpler and more direct solution, the **third method**. Only draw lines from one validated GPS position to the other.

In this case as soon as the road have a bit of a curve, it will lead to lines "cutting" over places where there is no road:
![image](https://user-images.githubusercontent.com/6548184/128784360-9b84d983-bae0-485d-8173-dc8ed4246bce.png)

I can see a similar representation in the TxDOT feed currently being published, where the workzone is on the I-635 but the LineString representation goes straight between two points:
![image](https://user-images.githubusercontent.com/6548184/128784568-392e9ac7-563b-4567-b95b-88fdcf2ce998.png)

Is this an acceptable output?
Of course, it depends on the feed consumers, and where they plan to display that geometry information. In some situations it could be accepted, but I believe on a public website this wouldn't look good, and for an automated vehicle it could be dangerous..

 ## Proposed changes

Unfortunately, we are better just to stop trying to publish a path that can't be confirmed to be correct. Since LineString geometry representation are kinda impossible to be attained automatically based solely on roadside devices GPS positions with a 100% certainty and following road curvatures, I believe MultiPoint would be a better fit in this case.

This lead to my proposed change. Remove the line from above description "MultiPoint should be used when only the start and end coordinates are known.". I would also remove the rule requiring MultiPoint to be only composed of pair and would allow a single RoadEvent to be composed of as many MultiPoint as possible, as long as they are representing the same workzone event.

The reason being that if you have a workzone with hundreds of sensor points, but the same information for the RoadEvent composing the whole workzone, it would be a bit irrational to split the workzone into hundreds of small events composed of start/stop pairs while the rest of the information remains the same and get duplicated leading to a huge output. It would be much better to have a single RoadEvent entry with the hundreds of points representing valid positions on the workzone, without any path between them.

**Comment by j-d-b** [comment, marked as answer]
@Dunge thank you for posting this. I'm going to move this to the "Issues" section as I think it should be addressed by changes to the specification, as you proposed.

**Reply by j-d-b** [reply]
See #194.

**Comment by rdsheckler** [comment]
Properly addressing this issue requires the group to decide if we are going to have entries in the WZDX that do not have direct human input or not.  If we have people entering geometry for a WZDX feed we can put a lot of structure into the requirements of the geometry definition.  If we are relying on gps located equipment to make automatic WZDX entries we will need to recognize that very simple geometric representations will need to be allowed.  Most work zones will be represented by a first point and nothing else.  Either case is fine, precision, complicated geometry that has human input or automated systems that may only have one or two points.  But, we do need to decide if we want more entries that are simpler or fewer entries that have more detail.

As a point of reference, our firm is reporting between 4,000 and 5,000 unique events each day.  Most of those are a single point where the work starts.  We haven't found a way to get the crews to do much more than give us that first point.  As a group we need to decide if that is an entry that is worthwhile before we finalize the subject of geometric representation.

**Reply by j-d-b** [reply]
@rdsheckler one of the goals of WZDx is to get more work zone data to the public. That means the least barriers to entry that is reasonable (minimal number of required fields, flexibility for content of properties). Direct human input should not be required to create a WZDx feed. I think more simpler entries meets the goal of WZDx.

The single start point only is technically allowed in the WZDx specification right now, as the producer can use MultiPoint and provide the same point for the start and end, and give the `ending_accuracy` as `estimated`. If this is a common case, allowing a single Point Geometry would be beneficial. Please see #194 for a continuation of the discussion of the changes we can make to the WZDx specification.
